### ISSUE TITLE
Lean mode freezing Emacs and sluggish VSCode 

### ISSUE THREAD

**Issue by leodemoura** [leanprover/lean3#1364]
As @rlewis1988 also reported, Emacs often freezes when we are editing the standard library.
Yesterday, I had to disable lean-mode to be able to edit `tactic.lean`, otherwise I would have to wait 5 secs after each keystroke.

I built a repro the exposes the problem. To execute the repro, we need to use the latest version which adds the command `sleep` and `long_sleep` to the Lean server.

As far as I can tell, the problem happens when we open files `A.lean` and `B.lean`, where `B.lean` depends on `A.lean` (and many other files that also depend on `A.lean`). Then, we start editing `A.lean` (e.g., introducing an fixing errors).

In my first repro: https://gist.github.com/leodemoura/0e3be8b2f2f364e0677392db9e8f2dc7

I open `tactic.lean`,  open `vector.lean`, break the `tactic_result` declaration at `tactic.lean`, and then fix it.
For this sequence of operations, the lean server produces a 210Mb output trace. This trace corresponds to 2 keystrokes. I imagine that yesterday when I was editing the file, several terabytes of information were being sent between Emacs and Lean server.

By inspecting the trace, it seems the server is producing longer and longer messages `{"msg", ...}`, where message `i` is a prefix of message `i+1`.

In the second repro: https://gist.github.com/leodemoura/d9e78ba6f2fffdbe0400344c3d6f10ee

The sequence of events is similar, but I do not open `vector.lean`. In this case, the lean server produces a 5Mb trace

This problem is not specific to the standard library. It seems it will affect any project that has many files, and non trivial dependencies. 

I also tried the experiment with VS Code. It doesn't freeze, but everything is very slow, and the feedback produced by the editor is more or less useless since it takes a long time to stabilize. 

@rlewis1988: I'm using the following workaround to edit stdlib files, I open a fresh Emacs process, and open only the file I want to edit.

**Comment by gebner** [comment]
Thanks for making an easily reproducible example.  Not that for reproduction you need to replace `\/home\/leo\/projects\/lean` with your local lean checkout.

In this case the server produces 3857904 messages.  Many of those are full message updates, that is, where we send all available messages again, because an old message got removed.  I guess we can just limit these full refreshes to once every 200ms or so.

**Comment by gebner** [comment]
Just some more data.  Virtually all of the messages are full refreshes.  Limiting them to once every 100ms would cut down the responses by a factor of over 8000!
```
$ lean --server=tactic_vector_trace.in | grep all_messages | wc
   1515 3850187 202859989
```

**Comment by leodemoura** [comment]
> Many of those are full message updates, that is, where we send all available messages again, because an old message got removed.

This explains the behavior. I was confused because the "tsunami of messages" was being produced after I fixed the typo at `tactic.lean`.

> I guess we can just limit these full refreshes to once every 200ms or so.

Yes, I agree.

**Comment by robertylewis** [comment]
> @rlewis1988: I'm using the following workaround to edit stdlib files, I open a fresh Emacs process, and open only the file I want to edit.

I've been doing similar to edit library_dev. This change would be nice!

**Comment by gebner** [comment]
Sorry, I failed at reading:
```
$ lean --server=tactic_vector_trace.in | grep all_messages | wc
   1515 3850187 202859989
```
That's just 1515 full updates, which is much more reasonable, but still at 40Hz.

**Comment by leodemoura** [comment]
@gebner The commit above reduced the amount of data communicated, but Emacs is still non responsive. 

The repro is:
1- open `library/init/meta/tactic.lean`
2- introduce a typo at line 30:
```
| success      : α → tactic_state → tactic_result
```
Example remove a `c` from `success`.

Any ideas on what is going on? Do you think the problem is emacs parsing the output?

**Comment by leodemoura** [comment]
BTW, now it is non responsive even if `tactic.lean` is the only file opened.

**Comment by gebner** [comment]
This is interesting.  I have just changed `success` to `succes`, now there are 236 errors and 52 warnings.  Even if I don't change anything anymore, emacs is terribly sluggish.

**Comment by gebner** [comment]
I think I have found the reason.  The task/progress notification is coupled to the messages.  Whenever the server status changes (processing vs. finished), we reload all messages.
This wouldn't be too bad, but every info request triggers a task, which triggers the progress notification.  I'll fix this tomorrow.

**Comment by leodemoura** [comment]
BTW, the emacs profiler claims that 97% of the runtime is being spent in the following function
```
(defun lean-server-show-messages (&optional buf)
  (with-current-buffer (or buf (current-buffer))
    (lean-server-update-task-overlays)
    (flycheck-buffer)))  <<<<<< HERE
```

**Comment by gebner** [comment]
Okay, I've got some good news and some bad news.  Good news:
 * Emacs no longer hangs if you move around the file *without changing* anything.
 * VS Code copes pretty well with changing `tactic.lean` while `standard.lean` is open.

Bad news:
 * Emacs still freezes when changing `tactic.lean` while `standard.lean` is open.  (Half of the time is spent updating flycheck errors, 30% on parsing JSON. Thanks for telling me about the profiler, BTW!)
 * There is a small memory leak, we leak about 200M per success->succes->success cycle.  (Fortunately we can override the memory limit via customize.)

I wanted to refactor the whole parallel compilation infrastructure a bit anyhow, I think I now have a few high-level goals for that:
 * Message updates need to be a bit more fine-grained than "something changed, we resend everything".  Just increasing the granularity to the file level would probably solve most of the performance issues with many files.  I'm not sure whether we should track individual messages ("messages 34, 37, and 43 got deleted; here is message 111"), this seems complicated and I'm not sure if flycheck/vscode actually support these operations.
 * Reverse dependencies must not be processed immediately.  If changing a line in `tactic.lean` did not cause all reverse dependencies to be recompiled (and many error messages to be changed), then we would not have this problem either.  My original plan was to delay the compilation of the reverse dependencies by a few seconds.
 * We absolutely need information about the currently visible files and probably even the currently visible lines in the server.  Then we could do Isabelle-style "only compile to the bottom of the screen".  We could also do "compile only the current screen", and then add an extra "compile the whole project" command.
 * The processing status should not be coupled to the existence of running tasks.  The info command should not change the status icon to hourglass/refresh.

Any other ideas or wishes?

**Comment by Kha** [comment]
> I'm not sure whether we should track individual messages ("messages 34, 37, and 43 got deleted; here is message 111"), this seems complicated and I'm not sure if flycheck/vscode actually support these operations.

Semi-related, I was thinking about about tracking individual timeouts per message on the server side. If we only pushed out messages that have not been resolved within, say, 300ms, it should make the flycheck buffer much less jittery. The same logic may work for the status icon and especially the background coloring.

**Comment by leodemoura** [comment]
> There is a small memory leak, we leak about 200M per success->succes->success cycle. (Fortunately we can override the memory limit via customize.)

Which one? I'm only aware of the following memory leaks:

```
==24086== 168 (40 direct, 128 indirect) bytes in 1 blocks are definitely lost in loss record 10 of 12
==24086==    at 0x4C2D1AF: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==24086==    by 0x380D09: lean::mt_task_queue::spawn_worker() (in /home/leo/projects/lean/bin/lean)
==24086==    by 0x383677: lean::mt_task_queue::wait(lean::generic_task_result const&) (in /home/leo/projects/lean/bin/lean)
==24086==    by 0x2F434B: lean::import_module(lean::environment&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, lean::module_name const&, std::function<std::shared_ptr<lean::loaded_module const> (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, lean::module_name const&)> const&, lean::buffer<lean::import_error, 16u>&) (in /home/leo/projects/lean/bin/lean)
==24086==    by 0x2F9193: lean::mk_preimported_module(lean::environment const&, lean::loaded_module const&, std::function<std::shared_ptr<lean::loaded_module const> (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, lean::module_name const&)> const&) (in /home/leo/projects/lean/bin/lean)
==24086==    by 0x2F92EA: lean::preimport_task::execute() (in /home/leo/projects/lean/bin/lean)
==24086==    by 0x2F655E: lean::task_result_cell<lean::environment>::execute_and_store_result() (in /home/leo/projects/lean/bin/lean)
==24086==    by 0x385E68: lean::execute_task_with_scopes(lean::generic_task_result_cell*) (in /home/leo/projects/lean/bin/lean)
==24086==    by 0x3830E0: std::_Function_handler<void (), lean::mt_task_queue::spawn_worker()::{lambda()#1}>::_M_invoke(std::_Any_data const&) (in /home/leo/projects/lean/bin/lean)
==24086==    by 0x2AADED: lean::lthread::imp::_main(void*) (in /home/leo/projects/lean/bin/lean)
==24086==    by 0x5DD1709: start_thread (pthread_create.c:333)
==24086==    by 0x60F00AE: clone (clone.S:105)
==24086== 
==24086== 672 (160 direct, 512 indirect) bytes in 4 blocks are definitely lost in loss record 12 of 12
==24086==    at 0x4C2D1AF: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==24086==    by 0x380D09: lean::mt_task_queue::spawn_worker() (in /home/leo/projects/lean/bin/lean)
==24086==    by 0x3811F7: lean::mt_task_queue::mt_task_queue(unsigned int, std::function<lean::task_priority (lean::generic_task*)> const&) (in /home/leo/projects/lean/bin/lean)
==24086==    by 0x381498: lean::mt_task_queue::mt_task_queue(unsigned int) (in /home/leo/projects/lean/bin/lean)
==24086==    by 0x26E15B: main (in /home/leo/projects/lean/bin/lean)
```

**Comment by leodemoura** [comment]
> We absolutely need information about the currently visible files and probably even the currently visible lines in the server. 

+1000 for this one.

BTW, in the `tactic.lean` and `standard.lean` scenario, we are saving a lot of of useless information in the `info_manager`.  It is saving `info_data` for all reached files (a big chunk of the stdlib), but only two of them are visible. 

Here is some data for the `tactic.lean` + `standard.lean` scenario (with `-j 2`): open both, introduce bug in `tactic.lean`, and then fix it.

- With info: 490 Mb
- Without info: 392 Mb

So, a big chunk (100Mb) is being consumed by the `info_manager`.
The overhead for snapshots is not that big. If we disable snapshots we save only 12Mb.
It makes since we do not store snapshots at `module_mgr::get_direct_imports`.

**Comment by leodemoura** [comment]
Here is memory consumption data for compiling the standard library using `--make`.

- `-j 0` 180Mb
- `-j 1` 309Mb
- `-j 2` 319Mb
- `-j 4` 331Mb
- `-j 8` 367Mb

**Comment by leodemoura** [comment]
I'm experiencing instabilities after I merged #1367 
The squiggly marks (green and red) are not being updated correctly. I have to restart the server by pressing `Ctrl-c-Ctrl-r` to fix the problem. I can reproduce the problem consistently by opening some random .lean file, and making a change.
Here is a sequence
1- open `library/init/data/div.lean`
2- remove `t` from `nat` at `instance : has_div nat :=`.
3- put the `t` back.
Now, I have a squiggly green line at `private def div.F (x : nat) (f : Π x₁, x₁ < x → nat → nat) (y : nat) : nat :=`.
It goes away if I restart the server or 
- Switch focus to another Windows, and then return to Emacs.

**Comment by gebner** [comment]
> If we only pushed out messages that have not been resolved within, say, 300ms, it should make the flycheck buffer much less jittery.

At the moment, we only push out message updates every 100ms.  We still have a timer on the emacs side: when flycheck takes more than 100ms to render the messages, then we can skip updates.

> Which one? I'm only aware of the following memory leaks:

This was just an observation with `top`.  I just observed that the memory use increased by 200M every time I changed `tactic.lean`.  Maybe it's just the thread-local caches being populated and the memory usage would stabilize after some time.  I'll investigate this.

> BTW, in the tactic.lean and standard.lean scenario, we are saving a lot of of useless information in the info_manager. [...] With info: 490 Mb, Without info: 392 Mb

20% overhead is not too bad, I'd take it over having to recompile the files for info_manager support.

> Here is memory consumption data for compiling the standard library using --make.  -j 0 180Mb, -j 1 309Mb, ..., -j 8 367Mb

I'm surprised at the big jump from single-threaded to multi-threaded with one thread.  Maybe it's because with -j1 we already spawn more than one thread: the noncomputability check introduces a synchronization point with a few lemmas, where we actually need to look at the proof.  In those cases we spawn an extra thread to elaborate and check the proof (since the first thread is already occupied with parsing).

**Comment by gebner** [comment]
> Now, I have a squiggly green line at private def div.F (x : nat) (f : Π x₁, x₁ < x → nat → nat) (y : nat) : nat :=.

I see the problem: the green line indicates a running task.  Since we no longer update flycheck on task changes, these flycheck messages no longer get updated.

**Comment by robertylewis** [comment]
@leodemoura Have the sluggishness problems you were seeing gone away completely?

Last night I updated library_dev to compile with the current Lean. To experiment, I opened all five files in data/list. I noticed the following:
- Occasionally, fixing an error would reset all Flyspeck error messages -- the counter would go to 0 and then build back up. Emacs was unusable while it did this, which took a few seconds. Oddly, I couldn't consistently reproduce this. Recreating the error and fixing it again usually wouldn't trigger the same behavior.
- This morning, I pulled the new version of Lean and ran `lean --make` in the library_dev folder. (Everything besides auto compiles.) With all data/list files open, looking at sort.lean, I restarted lean-server. Flycheck gave me a checkmark almost immediately, and claimed 0/0 messages, but over the course of about 30 seconds the message count changed, eventually to 2/6. The two errors are `(deterministic) timeout` errors, including
```
theorem sorted_merge_sort : ∀ l : list α, sorted r (merge_sort l) :=
sorry
```
which I wouldn't expect to time out. As I copied that theorem statement, the error count jumped to 3 -- there is now a `(deterministic) timeout` error reported at the first character of the file.
- I closed emacs, removed the .olean files, and ran `lean --make` again. Everything compiled quickly (no timeout errors). Then I opened basic, sort, and set. Flycheck now showed the hourglass when compiling sort.lean, but took at least 20 seconds and still reported timeout errors. I switched to set.lean, which reported a dead lean-server; I restarted the server, switched back to sort.lean, and Flycheck reported a dead lean-server there. This behavior is reproducible. 

I'm not sure how much of this is related to the original issue. Presumably the first point is, at least. Can anyone reproduce that by starting to edit the list files in the current, broken library_dev?

**Comment by robertylewis** [comment]
> I closed emacs, removed the .olean files, and ran lean --make again. Everything compiled quickly (no timeout errors). 

Actually, I take this back -- I just tried again, and it does take a while, but does eventually succeed without any unexpected messages.

**Comment by leodemoura** [comment]
@rlewis1988 My editor has been freezing, but from time to time, I get a bunch of red squiggly marks because the server is out of memory. 

@gebner Any ideas on how to deal with this? Is there an easy way to reclaim memory?
BTW, I'm not sure where the memory is being spent in sever mode.
You mentioned a memory leak. Did you investigate this issue?

We don't have a server-friendly cache datastructure. We never remove old entries from the cache. 
We only reclaim memory when the cache is flushed.
If the problem is the thread local caches, I can fix it by using a smarter data-structure that removes old entries.

**Comment by gebner** [comment]
> You mentioned a memory leak. Did you investigate this issue?

Not yet, I am a bit busy right now (not only with Lean), and also a bit sick.  But it is on my to-do list.
It is possible that this leak was just the expression cache that we disabled now.

> We don't have a server-friendly cache datastructure. We never remove old entries from the cache.

Never, as in we don't even have a maximum?  In some caches (like replace_cache) we actually have a maximum size, maybe we should do that everywhere?

Would it help if we killed worker threads when we no longer need them?  (If a task needs to wait on another task, we spawn a replacement worker thread.  These replacements threads never die, but we reuse them.  When compiling the standard library, I think we spawn 3 replacements threads.)

We could also try to kill all worker threads after some time of inactivity.  But this might have a negative impact on latency.

In any case I'm not so worried about using *much* memory, as long as the memory usage does not grow over time.

**Comment by leodemoura** [comment]
> Never, as in we don't even have a maximum?

No, they are regular hashtables.

> In some caches (like replace_cache) we actually have a maximum size, maybe we should do that everywhere?

Yes, this is a good idea. We can try to use the approach used at `replace_cache` in all other caches.
BTW, `replace_cache` is an exception. Most of our caches are regular hashtables. 

> Would it help if we killed worker threads when we no longer need them? 

I think so. We would reclaim all thread local storage.
For example, suppose one of the threads process a complex lemma once.
The thread local memory pools and small object allocators will contain a lot of unused memory.

**Comment by gebner** [comment]
> I think so. We would reclaim all thread local storage.

In PR #1405, I have implemented this.  But even killing all worker threads (after an idle time of 1000ms) doesn't help.  Memory usage still steadily grows with (almost) every keystroke.  I'm kind of at a loss here.

**Comment by gebner** [comment]
Is this issue still present for you?  I have been running lean with jemalloc for a month now, and can't recall running into any memory problems.

The remaining memory problems seem to be due to fragmentation.  That is, we allocate, say, 700M, but then the allocator needs twice that much RAM.  This is not an issue with jemalloc where we can easily query the number of allocated bytes.  Thread-local caches should be cleared when idling since we kill the worker threads.

**Comment by gebner** [comment]
> Is this issue still present for you?

Sorry for nagging again.  Are there concrete problems left that we can solve?

Personally, everything runs pretty smoothly for me.  I haven't had out-of-memory errors for months.  But it's now happened several times that I just don't see a problem because I'm running emacs 25/vscode/whatever, so I thought I'd ask again.

**Comment by leodemoura** [comment]
Yes, things are running smoothly for me too.
Let's close this issue. If we observe problems, we can reopen it.
